tfe/data_source_outputs.go

```diff
@@ -0,0 +1,234 @@
+			"workspace":    tftypes.String,
+			"organization": tftypes.String,
+			"values":       tftypes.Object{AttributeTypes: stateTypes},
+			"id":           tftypes.String,
+		},
+	}, map[string]tftypes.Value{
+		"workspace":    tftypes.NewValue(tftypes.String, wsName),
+		"organization": tftypes.NewValue(tftypes.String, orgName),
+		"values":       tftypes.NewValue(tftypes.Object{AttributeTypes: stateTypes}, tftypesValues),
+		"id":           tftypes.NewValue(tftypes.String, id),
+	}))
+
+	if err != nil {
+		return &tfprotov5.ReadDataSourceResponse{
+			Diagnostics: []*tfprotov5.Diagnostic{
+				{
+					Severity: tfprotov5.DiagnosticSeverityError,
+					Summary:  "Error encoding state",
+					Detail:   fmt.Sprintf("Error encoding state: %s", err.Error()),
+				},
+			},
+		}, nil
+	}
+	return &tfprotov5.ReadDataSourceResponse{
+		State: &state,
+	}, nil
+}
+
+func (d dataSourceOutputs) ValidateDataSourceConfig(ctx context.Context, req *tfprotov5.ValidateDataSourceConfigRequest) (*tfprotov5.ValidateDataSourceConfigResponse, error) {
+	return &tfprotov5.ValidateDataSourceConfigResponse{}, nil
+}
+
+func (d dataSourceOutputs) readConfigValues(req *tfprotov5.ReadDataSourceRequest) (string, string, error) {
+	var orgName string
+	var wsName string
+	var err error
+
+	config := req.Config
+	val, err := config.Unmarshal(tftypes.Object{
+		AttributeTypes: map[string]tftypes.Type{
+			"workspace":    tftypes.String,
+			"organization": tftypes.String,
+			"values":       tftypes.DynamicPseudoType,
+			"id":           tftypes.String,
+		}})
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error unmarshalling config: %v", err)
+	}
+
+	var valMap map[string]tftypes.Value
+	err = val.As(&valMap)
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error assigning configuration attributes to map: %v", err)
+	}
+
+	if valMap["organization"].IsNull() || valMap["workspace"].IsNull() {
+		return orgName, wsName, fmt.Errorf("Organization and Workspace cannot be nil: %v", err)
+	}
+
+	err = valMap["organization"].As(&orgName)
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error assigning 'organization' value to string: %v", err)
+	}
+
+	err = valMap["workspace"].As(&wsName)
+	if err != nil {
+		return orgName, wsName, fmt.Errorf("Error assigning 'workspace' value to string: %v", err)
+	}
+
+	return orgName, wsName, nil
+}
+
+type stateData struct {
+	outputs map[string]*outputData
+}
+
+type outputData struct {
+	Value cty.Value
+}
+
+func (d dataSourceOutputs) readStateOutput(ctx context.Context, tfeClient *tfe.Client, orgName, wsName string) (*stateData, error) {
+	log.Printf("[DEBUG] Reading the Workspace %s in Organization %s", wsName, orgName)
+	ws, err := tfeClient.Workspaces.Read(ctx, orgName, wsName)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading workspace: %v", err)
+	}
+
+	log.Printf("[DEBUG] Reading the current StateVersion for Workspace ID %s.", ws.ID)
+	sv, err := tfeClient.StateVersions.Current(ctx, ws.ID)
```

Comment:
* This defeats the intent of not needing to be able to read state versions to read outputs 😅 
* You shouldn't need to call out for the current state version anyway, it's a needless call; a user without state version access should be able to fetch the ID from the relationships member in the workspace response and use that directly.

Reply:
Oops, this was an artifact of the last attempt, and will change it to use the workspace response. :sigh:, stale context.